wp_api/src/middleware.rs

```diff
@@ -0,0 +1,364 @@
+    where
+        Self: Sync,
+    {
+        async move {
+            let pipeline = &self.get_middleware_pipeline();
+            let response = self.get_request_executor().execute(request.clone()).await?;
+
+            pipeline
+                .process(
+                    self.get_request_executor().clone(),
+                    response,
+                    request.clone(),
+                )
+                .await
+                .map_err(|e| match e {
+                    WpApiError::RequestExecutionFailed {
+                        status_code,
+                        redirects,
+                        reason,
+                    } => RequestExecutionError::RequestExecutionFailed {
+                        status_code,
+                        redirects,
+                        reason,
+                    },
+                    _ => RequestExecutionError::RequestExecutionFailed {
+                        status_code: None,
+                        redirects: None,
+                        reason: RequestExecutionErrorReason::GenericError {
+                            error_message: e.to_string(),
+                        },
+                    },
+                })
+        }
+    }
+}
+
+// MARK: - RetryMiddleware
+
+#[derive(Debug, uniffi::Object)]
+struct RetryMiddleware {
+    max_retries: u32,
+    max_retry_wait_seconds: u64,
+}
+
+#[uniffi::export]
+impl RetryMiddleware {
+    #[uniffi::constructor]
+    fn new(max_retries: u32, max_retry_wait_seconds: u64) -> Self {
+        println!("Creating retry middleware");
+        Self {
+            max_retries,
+            max_retry_wait_seconds,
+        }
+    }
+}
+
+#[uniffi::export]
+#[async_trait::async_trait]
+impl WpApiMiddleware for RetryMiddleware {
+    async fn process(
+        &self,
+        request_executor: Arc<(dyn RequestExecutor + 'static)>,
+        response: WpNetworkResponse,
+        request: Arc<WpNetworkRequest>,
+    ) -> Result<WpNetworkResponse, WpApiError> {
+        let mut retry_count = 0;
+        let mut response = response;
+
+        loop {
+            if retry_count >= self.max_retries {
+                return Err(WpApiError::RequestExecutionFailed {
+                    status_code: Some(response.status_code),
+                    redirects: None,
+                    reason: RequestExecutionErrorReason::MisconfiguredRateLimitError {},
+                });
+            }
+
+            if response.is_rate_limit_exceeded() && self.max_retries != 0 {
+                let retry_after = response.get_retry_after();
+
+                if let Some(mut retry_after) = retry_after {
+                    // If the server sends some super-long value, we don't want to wait that long
+                    if retry_after > self.max_retry_wait_seconds {
+                        retry_after = self.max_retry_wait_seconds;
+                        async_sleep(Duration::from_secs(retry_after)).await;
+                    }
```

Comment:
This would only sleep if `retry_after` is longer than `max_retry_wait_seconds`. In my suggestion PR, I am changing this as:

```
request_executor.sleep(
    // If the server sends some super-long value, we don't want to wait that long
    Duration::from_secs(std::cmp::min(
        retry_after,
        self.max_retry_wait_seconds,
    ))
    .as_millis(),
);
```

Since this includes a logic change, I wanted to leave a note just in case I am misunderstanding something.

Reply:
Yeah, when I implemented `async_sleep`, I think I dropped the reference to it one line too high.